### PR TITLE
WT-13849 Fix read pinned_timestamp multiple times in __wt_txn_pinned_timestamp

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by ajmorton** [comment]
@quchenhao have we seen test failures based on this? The acquire barrier already guarantees we read `pinned_timestamp` before we read `checkpoint_timestamp`, and it prevents invented reads of `pinned_timestamp` from appearing after the barrier which gives us READ_ONCE semantics relative to `checkpoint_timestamp`.

If we're seeing issues it's worth reviewing where we write the pinned and checkpoint timestamps to make sure there's correct release semantics there as well

**Comment by quchenhao** [comment]
> @quchenhao have we seen test failures based on this? The acquire barrier already guarantees we read `pinned_timestamp` before we read `checkpoint_timestamp`, and it prevents invented reads of `pinned_timestamp` from appearing after the barrier which gives us READ_ONCE semantics relative to `checkpoint_timestamp`.
> 
> If we're seeing issues it's worth reviewing where we write the pinned and checkpoint timestamps to make sure there's correct release semantics there as well

No, we haven't seen any failure yet.

I think this code may already cause us to read twice. The following barrier doesn't matter.

*pinned_tsp = pinned_ts = txn_global->pinned_timestamp;

File: src/txn/txn_timestamp.c [hunk 1]
```diff
@@ -304,10 +304,10 @@ __wti_txn_update_pinned_timestamp(WT_SESSION_IMPL *session, bool force)
     __wti_txn_get_pinned_timestamp(
       session, &pinned_timestamp, WT_TXN_TS_ALREADY_LOCKED | WT_TXN_TS_INCLUDE_OLDEST);
 
-    if (pinned_timestamp != 0 &&
+    if (pinned_timestamp != WT_TS_NONE &&
       (!txn_global->has_pinned_timestamp || force ||
         txn_global->pinned_timestamp < pinned_timestamp)) {
-        txn_global->pinned_timestamp = pinned_timestamp;
+        WT_RELEASE_WRITE_WITH_BARRIER(txn_global->pinned_timestamp, pinned_timestamp);
         txn_global->has_pinned_timestamp = true;
```
Comment (quchenhao) [inline src/txn/txn_timestamp.c:311]: @ajmorton I think the has_pinned_timestamp write is also not correct. It is not defined as a shared variable, which is wrong. I don't want to fix it here. May be we can create a separate ticket.

File: src/include/txn_inline.h [hunk 2]
```diff
@@ -842,11 +849,12 @@ __wt_txn_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_tsp)
      * pinned. If a checkpoint is starting and we have to use the checkpoint timestamp, we take the
      * minimum of it with the oldest timestamp, which is what we want.
      */
-    WT_ACQUIRE_BARRIER();
     checkpoint_ts = txn_global->checkpoint_timestamp;
```
Comment (quchenhao) [inline src/include/txn_inline.h:852]: @ajmorton  Not sure why there is no warnings for this read. checkpoint_timestamp is a shared variable and should be read using atomic read.

File: src/txn/txn_timestamp.c (hunk 1 above)
Comment (ajmorton) [reply src/txn/txn_timestamp.c:311]: Yep that's likely. txn_global was scoped out of PM-3221 due to complexity and is scheduled for [SPM-3426](https://jira.mongodb.org/browse/SPM-3426). [WT-11149](https://jira.mongodb.org/browse/WT-11149) is probably the ticket for this.

Comment (ajmorton) [reply src/txn/txn_timestamp.c:311]: Can you move the `WT_RELEASE_WRITE_WITH_BARRIER` barrier onto `txn_global->has_pinned_timestamp`? 
`WT_RELEASE_WRITE_WITH_BARRIER` is defined as
```
#define WT_RELEASE_WRITE_WITH_BARRIER(v, val)   \
    do {                                        \
        WT_RELEASE_BARRIER();                   \
        __wt_atomic_store_generic(&(v), (val)); \
    } while (0)
```
so the current implementation is doing
```
RELEASE_BARRIER()
txn_global->pinned_timestamp = pinned_timestamp;
txn_global->has_pinned_timestamp = true
```
and not enforcing any ordering between the two writes. 
Similarly for the `WT_ACQUIRE_READ_WITH_BARRIER` call.

File: src/include/txn_inline.h (hunk 2 above)
Comment (ajmorton) [reply src/include/txn_inline.h:852]: Annoyingly there's no way to check this. We could if we moved to _Atomic but MSVC support is still lacking. For now we're relying on developers to use the shared vars correctly.

File: src/txn/txn_timestamp.c (hunk 1 above)
Comment (quchenhao) [reply src/txn/txn_timestamp.c:311]: I think we need both not only on the has_pinned_timestamp variable. But I don't want to increase the scope of this change so my initial idea is just to enforce we only read pinned_timestamp once.

File: src/include/txn_inline.h (hunk 2 above)
Comment (quchenhao) [reply src/include/txn_inline.h:852]: Do you want me to fix this in the pr?

**Comment by quchenhao** [comment]
Here's the bug case I think can happen:

```
1) pinned_ts = txn_global->pinned_timestamp;           // ts = 100
2)  *pinned_tsp = txn_global->pinned_timestamp;      // ts = 120

3) // acquire barrier here

4) checkpoint_ts = txn_global->checkpoint_timestamp; // ts = 110
5) if(checkpoint_ts < pinned_ts)                     // 110 > 100
6)    pass                                           // do nothing, we've already written *pinned_tsp == 120 above
```

File: src/include/txn_inline.h (hunk 2 above)
Comment (ajmorton) [reply src/include/txn_inline.h:852]: No that's ok. txn_global is going to need a much larger rework in a future project. I don't think we need to add noise to this PR.

File: src/txn/txn_timestamp.c (hunk 1 above)
Comment (ajmorton) [reply src/txn/txn_timestamp.c:311]: Agreed on `has_pinned_timestamp`. This is a clear case of message passing so I'd be happy to add the change in this PR. Perhaps we only add the ACQUIRE/RELEASE ops in `__wti_txn_update_pinned_timestamp` and `__wt_txn_pinned_timestamp` to keep the scope small? We can do the rest in the ticket to fix txn_global

File: src/include/txn_inline.h (hunk 2 above)
Comment (quchenhao) [reply src/include/txn_inline.h:854]: Thanks

File: src/txn/txn_timestamp.c (hunk 1 above)
Comment (quchenhao) [reply src/txn/txn_timestamp.c]: done

**Comment by quchenhao** [comment]
@ajmorton @kommiharibabu Can you review again?

**Review by ajmorton — APPROVED** [review]
LGTM 👍

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Pull request merged**